Ensure nStatus is set properly for all invalid blocks

jamesob opened the pull request:
- [x] **Prerequisite PR**: #12431 (it wouldn't be catastrophic to merge this before that, but preferable to get that in first.)

In trying to fix the fork warning code (`validation.cpp:CheckForkWarningConditions*`), it became apparent that we are marking some (but not all) invalid blocks as invalid (via `nStatus`) when they are received and subsequently dropped. The fact that we never mark some invalid blocks as such prevents us from e.g. detecting and warning on invalid chains with significant work.

This change has `ProcessNewBlock` call `InvalidateBlock` on the invalid block to do the expected bookkeeping in `mapBlockIndex` before dropping the block.

This change also consolidates the setting of CBlockIndex's `nStatus |= BLOCK_FAILED_VALID` to a single function (`InvalidBlockFound`) since there's peripheral bookkeeping (e.g. `g_failed_blocks.insert()`) that we want to do consistently but is duplicated in some places or not done in other cases when it apparently should be. 

[One such replacement with InvalidBlockFound](https://github.com/bitcoin/bitcoin/pull/12407/files?diff=unified#diff-24efdb00bfbe56b140fb006b562cc70bL3420) ensures addition of invalid blocks to `g_failed_blocks` and so (in theory) reduces CPU burden when being spammed with descendants of an invalid block, since we no longer have to walk `mapBlockIndex` to determine its invalidity. Based on reading usage of `g_failed_blocks` I can't tell if this savings is real, but in any case it seems worthwhile to keep that set consistent (i.e. 1-1 with blocks marked BLOCK_FAILED_VALID since last restart).

jamesob committed:
[tests] Add utility for generating bad coinbase transactions

Also adds a small logging clarification for `sync_blocks`. Includes feedback
from @ryanofsky.

jamesob committed:
[tests] Add a failing test that asserts invalidity of DoS-responding blocks

Ensure that we mark any block which merits a DoS response as invalid, as well
as its associated chain tip. This will become important for detecting and
warning about long invalid forks.

jamesob commented:
Hm: this changeset is failing tests that initially transmit a block with unexpected witness data (which gets rejected) and then retransmit without witness data for expected acceptance, but we get hung up on an existing `mapBlockIndex` entry that's been marked invalid.

Is it worth special-casing out `"unexpected-witness"` rejections from the `InvalidateBlock` call? Seems a little ugly.

Is the bookkeeping this PR wants to do even worth it? I guess we could just lump long invalid forks in with long headers-only forks, but that doesn't seem great either...

sipa commented:
You can only mark a block as invalid when fPossibleCorruption is not set in the validation status.

jamesob commented:
Ah good point, thanks @sipa. That seems to have fixed it.

dcousens reviewed (approved):
utACK 7bd2d85906f12661efb9e0747d8bbd30530d149d

jamesob commented:
I've rebased this changeset, removing the code that's now in #12431 as well as the semi-related log statement. I've also DRY'd up parts of validation which set `BLOCK_FAILED_VALID` on `nStatus` to use `InvalidBlockFound` instead.

ryanofsky commented on test/functional/test_framework/blocktools.py:
```diff
@@ -110,6 +112,16 @@ def create_transaction(prevtx, n, sig, value, scriptPubKey=CScript()):
+
+def create_coinbase_with_bad_txin(*args, **kwargs):
+    """Return a coinbase CTransaction with an invalid txin."""
+    kwargs['txin'] = CTxIn(
```
In commit "[tests] Add utility for generating bad coinbase transactions"

Could drop kwargs assignment with

```python
return create_coinbase(*args, **kwargs. txin=CTxIn(...))
```

jamesob replied:
Ah, Python 3.4 doesn't support this syntax (which is what we run on Travis). Will have to revert this.

ryanofsky commented on src/validation.cpp:
```diff
@@ -3410,10 +3419,21 @@ bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<cons
+            // here. It *does* happen within AcceptBlock below.
+            BlockMap::iterator it = mapBlockIndex.find(pblock->GetHash());
+
+            // Duplicate call to CorruptionPossible() here just to be careful.
```
In commit "Do the proper InvalidateBlock bookkeeping for DoSy blocks"

Duplicating the CorruptionPossible check here seems unnecessarily confusing and fragile to me, though perhaps there's an advantage I'm not seeing.

ryanofsky reviewed (commented):
Slight utACK 6af17e40890651e5e5f5968a1a131c6a48ce998c. Slight because I don't know all the implications of adding the new InvalidBlockFound calls.

New test code looks great.

I'd find the main commit ("Do the proper InvalidateBlock bookkeeping for DoSy blocks") easier to understand if it were split up, with the actual behavior change in ProcessNewBlock in one commit, and all the documentation/assertlock/DRY cleanup stuff in a different commit or set of commits.

jamesob committed:
Do the proper InvalidateBlock bookkeeping for DoSy blocks

When we receive a block which causes a DoS response, ensure that we mark its
status as "invalid" in `mapBlockIndex`. This is necessary for detecting and
warning about forks which are invalid but long.

jamesob committed:
Refactoring/DRYing for InvalidBlockFound

Consolidate BLOCK_FAILED_VALID nStatus updates to use InvalidBlockFound. Clean
up InvalidBlockFound for a defensive early exit if called when corruption is
possible.

jamesob reviewed (commented):
@ryanofsky thanks very much for the review and good feedback. I've incorporated your recommendations and split out the last commit into two: one for the actual behavioral change and one for the cleanup.

jamesob commented on src/validation.cpp:
```diff
@@ -2727,7 +2741,6 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
         it++;
     }
 
-    InvalidChainFound(pindex);
```
This now happens in the `InvalidBlockFound` call above.

MarcoFalke commented:
Needs rebase to fix travis (sorry)

jamesob commented:
Rebased.

ryanofsky commented:
Slight re-utACK a12ce0dde462911bb702070d31ed28ba32353a05. Again someone more knowledgeable than me should review safety of adding new `InvalidBlockFound` calls in `ProcessNewBlock`,  `AcceptBlock`, and `InvalidateBlock`.

Changes since last review were rebase, python cleanups, adding CorruptionPossible log print, and some suggestions from the last review.

You may want to update PR description since it will become part of git history if this is merged.

sipa commented on src/validation.cpp:
```diff
@@ -1277,14 +1277,27 @@ void static InvalidChainFound(CBlockIndex* pindexNew)
+    AssertLockHeld(cs_main);
+
+    if (state.CorruptionPossible()) {
+        LogPrintf("%s: can't be called when corruption is possible (invalid block=%s)",
```
Why log this unconditionally? I think it may be trivial to cause someone's logfile to be filled with these.

jamesob replied:
Yep, that's a good point. I'll change this into a comment.

jamesob commented:
Thanks @ryanofsky and @sipa for the re-review. I've incorporated @sipa's change (remove unconditional log in `InvalidBlockFound` when corruption is possible), rebased, and updated the PR description to better reflect the change's contents. I've added some text discussing the refactoring, reproduced here

> This change also consolidates the setting of CBlockIndex's `nStatus |= BLOCK_FAILED_VALID` to a single function (`InvalidBlockFound`) since there's peripheral bookkeeping (e.g. `g_failed_blocks.insert()`) that we want to do consistently but is duplicated in some places or not done in other cases when it apparently should be. 
> 
> [One such replacement with InvalidBlockFound](https://github.com/bitcoin/bitcoin/pull/12407/files?diff=unified#diff-24efdb00bfbe56b140fb006b562cc70bL3420) ensures addition of invalid blocks to `g_failed_blocks` and so (in theory) reduces CPU burden when being spammed with descendants of an invalid block, since we no longer have to walk `mapBlockIndex` to determine its invalidity. Based on reading usage of `g_failed_blocks` I can't tell if this savings is real, but in any case it seems worthwhile to keep that set consistent (i.e. 1-1 with blocks marked BLOCK_FAILED_VALID since last restart).

jamesob commented:
Any hope of a concept ACK/NACK on this from @sipa @TheBlueMatt @sdaftuar?

laanwj commented on src/validation.cpp:
```diff
@@ -2724,6 +2736,11 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
     return g_chainstate.PreciousBlock(state, params, pindex);
 }
 
+/**
```
The usual convention for method documentation in this project is to add the comment in the class definition, not above the implementation. This makes it easier to read the API documentation in one place when looking at the class.
That said, as both are in the `.cpp` file for this class I don't think that's much of an issue here.

jamesob replied:
Fixed, thanks.

jamesob commented:
Rebased and addressed @laanwj's feedback.

jamesob commented:
Closing this; isn't getting much traction and no point in rebasing if that's the case.
